File: benchmarks/comparisons/nimkdl/edit.nim
## Edit-then-encode throughput. Both nimkdl and kdl-rs are explicitly
## designed for byte-lossless edit-then-emit workflows (the editor /
## formatter / config-rewriter use case). This bench exercises the
## full cycle: parse(preserveFormat) → mutate one node → encode().
##
## nimkdl strategy: per-node `parseHash`; on encode, mutated subtrees
## emit canonical while siblings preserve verbatim source bytes.
##
## kdl-rs strategy: per-token whitespace storage; on encode, walks the
## tree emitting carried trivia per token regardless of mutation.
##
## The bench measures the COMPLETE cycle (parse + edit + encode) per
## iteration because that's the realistic editor "open → edit → save"
## workflow on a single file. Same cycle shape on both parsers.

import std/[monotimes, os, strformat, times]
import kdl

proc main() =
  if paramCount() < 1:
    echo "usage: nimkdl-edit <fixture-path>"
    quit(2)
  let path = paramStr(1)
  if not fileExists(path):
    echo &"missing fixture: {path}"
    quit(2)
  let src = readFile(path)

  const iters = 5_000
  # Warmup
  for _ in 0 ..< 100:
    var doc = parse(src, preserveFormat = true).get
    doc.nodes[0].setProp(doc, "bench-mark", newStringValue("edited"))
    discard encode(doc, emPreserve)
  let start = getMonoTime()
  for _ in 0 ..< iters:
    var doc = parse(src, preserveFormat = true).get
    doc.nodes[0].setProp(doc, "bench-mark", newStringValue("edited"))
    discard encode(doc, emPreserve)
  let el = (getMonoTime() - start).inNanoseconds.float / 1e9

  let us = el / float(iters) * 1_000_000.0
  let ops = float(iters) / el
  let fixture = path.extractFilename
  echo &"  nimkdl  edit-encode  {fixture:<30} {us:>8.1f}us avg   {ops/1000:>8.1f}K ops/s   {src.len} bytes"

main()

File: benchmarks/comparisons/kdl-rs/edit.rs
// Edit-then-encode bench for kdl-rs. Mirrors nimkdl/edit.nim:
// full cycle of parse → mutate one node → to_string per iteration.
// Matches the realistic "editor open → edit → save" workflow.
//
// kdl-rs strategy: per-token whitespace storage on every node; on
// to_string() the tree is walked emitting carried trivia per token
// regardless of which subtree was mutated.
use std::env;
use std::fs;
use std::path::Path;
use std::time::Instant;

fn main() {
    let args: Vec<String> = env::args().collect();
    if args.len() < 2 { eprintln!("usage: kdlrs-edit <fixture-path>"); std::process::exit(2); }
    let path = &args[1];
    if !Path::new(path).exists() {
        eprintln!("missing fixture: {}", path); std::process::exit(2);
    }
    let src = fs::read_to_string(path).expect("read fixture");

    let iters: u64 = 5_000;
    // Warmup
    for _ in 0..100 {
        let mut doc = kdl::KdlDocument::parse_v2(&src).expect("parse");
        {
            let n = &mut doc.nodes_mut()[0];
            let entry = kdl::KdlEntry::new_prop("bench-mark", "edited");
            n.push(entry);
        }
        let _ = doc.to_string();
    }
    let start = Instant::now();
    for _ in 0..iters {
        let mut doc = kdl::KdlDocument::parse_v2(&src).expect("parse");
        {
            let n = &mut doc.nodes_mut()[0];
            let entry = kdl::KdlEntry::new_prop("bench-mark", "edited");
            n.push(entry);
        }
        let _ = doc.to_string();
    }
    let el = start.elapsed().as_secs_f64();
    let us = el / iters as f64 * 1_000_000.0;
    let ops = iters as f64 / el;
    let fixture = Path::new(path).file_name()
        .map(|s| s.to_string_lossy().into_owned())
        .unwrap_or_else(|| "?".into());
    println!(
        "  kdl-rs  edit-encode  {fixture:<30} {us:>8.1}us avg   {:>8.1}K ops/s   {} bytes",
        ops / 1000.0, src.len()
    );
}
